src/rules/awaitPromiseRule.ts

```diff
@@ -54,27 +55,33 @@ function walk(ctx: Lint.WalkContext<Set<string>>, tc: ts.TypeChecker) {
     return ts.forEachChild(ctx.sourceFile, cb);
 
     function cb(node: ts.Node): void {
-        if (isAwaitExpression(node) && !couldBePromise(tc.getTypeAtLocation(node.expression))) {
+        if (isAwaitExpression(node) && !containsType(tc.getTypeAtLocation(node.expression), isPromiseType)) {
             ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+        } else if (isForOfStatement(node) && node.awaitModifier !== undefined &&
+                   !containsType(tc.getTypeAtLocation(node.expression), isAsyncIterable)) {
+            ctx.addFailureAtNode(node.expression, Rule.FAILURE_FOR_AWAIT_OF);
         }
         return ts.forEachChild(node, cb);
     }
 
-    function couldBePromise(type: ts.Type): boolean {
-        if (Lint.isTypeFlagSet(type, ts.TypeFlags.Any) || isPromiseType(type)) {
-            return true;
-        }
-
-        if (isUnionOrIntersectionType(type)) {
-            return type.types.some(couldBePromise);
-        }
-
-        const bases = type.getBaseTypes();
-        return bases !== undefined && bases.some(couldBePromise);
+    function isPromiseType(name: string) {
+        return promiseTypes.has(name);
     }
+}
 
-    function isPromiseType(type: ts.Type): boolean {
-        const { target } = type as ts.TypeReference;
-        return target !== undefined && target.symbol !== undefined && promiseTypes.has(target.symbol.name);
+function containsType(type: ts.Type, predicate: (name: string) => boolean): boolean {
```

Comment:
seems like a utility we might want in `language/utils.ts`, what do you think?

Reply:
Makes sense. We also use a similar function in `return-undefined` and probably in other rules, too.
We need to find a better abstraction than the one used here to make it work for other cases. Will do that in a follow-up PR.